src/fscacher/cache.py

```diff
@@ -192,12 +194,19 @@ def to_tuple(self):
 class DirFingerprint:
     def __init__(self):
         self.last_modified = None
-        self.tree_fprints = {}
+        self.hash_ords = None
 
     def add_file(self, path, fprint: FileFingerprint):
-        self.tree_fprints[path] = fprint
-        if self.last_modified is None or self.last_modified < fprint.mtime_ns:
+        fprint_hash = list(
```

Comment:
Why is it a list?

Reply:
ha -- I think I found the winner!!!  https://stackoverflow.com/a/29409299/1265472 takes advantage of `int.from_bytes`

```python
    def encrypt2(var, key, byteorder=sys.byteorder):
        key, var = key[:len(var)], var[:len(key)]
        int_var = int.from_bytes(var, byteorder)
        int_key = int.from_bytes(key, byteorder)
        int_enc = int_var ^ int_key
        return int_enc.to_bytes(len(var), byteorder)
```

<details>
<summary>seems to be all kosher and the same as more explicit version (but x10 faster on my sample case)</summary> 

```shell
In [15]: s2 = "b"*129; b2=s2.encode()

In [16]: s1 = "a"*128; b1=s1.encode()

In [17]: byte_xor(b1, b2) == encrypt2(b1, b2)
Out[17]: True

In [18]: %timeit encrypt2(b1, b2)
968 ns ± 0.345 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [19]: %timeit byte_xor(b1, b2)
10.3 µs ± 18.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

```
</details>

Reply:
the only thing I am not sure is trimming of one value based on the length of the other... (and in original zip implementations too if I got it right)

```
In [23]: encrypt2(b"123", b"12")
Out[23]: b'\x00\x00'

In [24]: encrypt2(b"12", b"123")
Out[24]: b'\x00\x00'
```

we might need first to get max len first and pad to avoid collisions

Reply:
<details>
<summary>here is a little script with removing trimming and a test randomizing the order and seems working Ok</summary> 

```python
import sys

def encrypt2(var, key, byteorder=sys.byteorder):
    int_var = int.from_bytes(var, byteorder)
    int_key = int.from_bytes(key, byteorder)
    int_enc = int_var ^ int_key
    return int_enc.to_bytes((int_enc.bit_length() + 7)//8, byteorder)

def xor_list(l):
    l = [s.encode() for s in l]
    if not l:
        return
    out = l[0]
    for v in l[1:]:
        out = encrypt2(out, v)
    return out

import random
random.seed(1)

l = ["abc", "a", "122", "", "al/s/1"]
#l = ["abc", "a", "122"] # , "", "al/s/1"]


r1 = xor_list(l)
print(r1)

# test that order does not matter
for i in range(100):
    random.shuffle(l)
    r = xor_list(l)
    if r1 != r:
        print(f"DIFF: {r}")
```
</details>

Reply:
obviously `xor_list` does not have to  be that ugly since there is reduce `functools.reduce(encrypt2, [v.encode() for v in l])`

Reply:
Here are the caching timings for a Zarr checksum calculation using this branch with the new bytes-XORing vs. the previous version vs. PR #70, all run against a directory containing 37480 files:

| fscacher version | Cache Miss | Cache Hit |
| --- | --- | --- |
| PR #70 | 63.4592 | 0.331767 |
| PR #71, previous | 67.1527 | 0.379186 |
| PR #71, current | 64.0528 | 0.328146 |

Reply:
Thank you for the timings @jwodder !  I am still really glad to see sub-second performance for "cache hit"s -- that is awesome!!!
as for #70 vs #71 (current) .  Probably impact of `sorted` is not as pronounced as I was afraid:

<details>
<summary>some crude timing shows that it is still just ~100ms for non-sorted list of 370k elements</summary> 

```ipython
In [2]: l = [str(i) for i in range(37000)]

In [3]: %timeit sorted(l)
614 µs ± 10.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [4]: random.seed(1); random.shuffle(l)

In [5]: %timeit sorted(l)
7.29 ms ± 151 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [6]: random.seed(2); random.shuffle(l)

In [7]: %timeit sorted(l)
7.43 ms ± 211 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [8]: l = [str(i) for i in range(370000)]

In [9]: %timeit sorted(l)
12.9 ms ± 167 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [10]: random.seed(1); random.shuffle(l)

In [11]: %timeit sorted(l)
120 ms ± 1.36 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
</details>

but I still feel that it is better to avoid `n*log(n)` operations (+ need to store long lists causing memory consumption) whenever possible, so I would keep this #71, current.